### PR TITLE
Typesafe and complete operators closes #580

### LINKED ISSUE #580: Filter Operators for DataStore
## Feature Request

The Operators currently available in DataStore doesn't contain all or enough of GraphQLCrud operators. 

These operators are also not typesafe.
For example; you can do this `p.anyField("gt", 5)` or `p.anyField("gt", "random string")`  in current DataStore without any compile time error.

Two things are wrong here;
- The "gt" operator should only be available on number fields
- The "gt" operator should only accept numbers as inputs

cc @wtrocki  @kingsleyzissou 

Edit:
See:
https://github.com/aerogear/graphback/blob/master/packages/graphback-core/src/runtime/QueryFilter.ts

https://github.com/aerogear/OpenVolunteerPlatform/blob/master/platform/client/src/dataFacade.tsx

### PR DESCRIPTION
**Pull request by Eunovo**
### Description

Implemented all the operators used in [GraphQLCrud filters](https://graphqlcrud.org/docs/next/find/#filtering). I modified Predicates such that the type of a field determines the allowed operators for that field. For example, numbers cannot use the "startsWith" operator. Also, only values of the same type or type[] can be compared. For example, you can perform the greater than operation on two numbers, not a number and a string.
 
##### Checklist

- [x] `npm test` passes
- [x] `npm run build` works
- [x] documentation is changed or added

### CODE REVIEW

File: packages/offix-datastore/src/predicates/Operators.ts [hunk 1]
```diff
@@ -1,10 +1,45 @@
-const equalsOperator = (modelField: any, value: any) => modelField === value;
-const gtOperator = (modelField: any, value: any) => modelField > value;
+type ArrayOperators = "ne" | "eq" | "in" | "contains";
```
Comment (wtrocki) [inline packages/offix-datastore/src/predicates/Operators.ts:1]: Once again. Would be best to use enums here?

File: packages/offix-datastore/src/predicates/Operators.ts [hunk 2]
```diff
@@ -1,10 +1,45 @@
...
+    T extends boolean[] ? ArrayOperators:
+    T extends boolean ? "ne" | "eq" | "in" :
+    T extends Date[] ? ArrayOperators:
+    T extends Date ? MathematicalOperators :
+    AllOperators;
+
+export const OperatorFunctionMap = {
+    ne: (m: any, v: any) => m !== v,
+    eq: (m: any, v: any) => m === v,
+    ge: (m: number | string | Date, v: number | string | Date) => m >= v,
```
Comment (wtrocki) [inline packages/offix-datastore/src/predicates/Operators.ts:37]: Really nice workaround for now. We can make this generic by accepting type itself. Something like here: https://github.com/aerogear/graphback/blob/master/packages/graphback-core/src/runtime/QueryFilter.ts#L82

**Comment by wtrocki** [comment]
@Eunovo how did you tested this knowing that current app is JS. Any trick I can do to check how typing are resolved in vscode?

**Comment by Eunovo** [comment]
Use the test files. DataStore.test.ts as a lot of predicates in it. You can check what types are resolved by playing with the code there

**Comment by wtrocki** [comment]
Actually enums here make less sense.
Screenshot 2020-07-09 at 15 59 40

**Comment by wtrocki** [comment]
Challenge here is that we cannot predict invalid predicate - like using int with string operation and throw error. I see this as good work but we definitely need follow up. Since we pass models we can actually add additional field into model that will check type and validate queries at runtime. 

WDYT @Eunovo

**Comment by Eunovo** [comment]
@wtrocki  don't merge yet. I wanna try something

**Comment by Eunovo** [comment]
> Challenge here is that we cannot predict invalid predicate - like using int with string operation and throw error. I see this as good work but we definitely need follow up. Since we pass models we can actually add additional field into model that will check type and validate queries at runtime.

@wtrocki We can pull this off but I'm thinking about the cost of doing this. Is runtime validation worth the effort? Do we need to verify predicates? What would happen if we didn't?

**Comment by wtrocki** [comment]
I mean this is stretch goal. What we have now it ok and we can improve this by generating typescript models or something - similar to generated schema typings from grpahql CRUD. I'm happy to merge this

**Comment by Eunovo** [comment]
@wtrocki I have fixed conflicts with master. You can merge

**Pull request merged**